presto-main/src/main/java/com/facebook/presto/operator/scalar/JsonToArrayCast.java

```diff
@@ -64,29 +70,36 @@ public ScalarFunctionImplementation specialize(BoundVariables boundVariables, in
     {
         checkArgument(arity == 1, "Expected arity to be 1");
         Type type = boundVariables.getTypeVariable("T");
-        Type arrayType = typeManager.getParameterizedType(StandardTypes.ARRAY, ImmutableList.of(TypeSignatureParameter.of(type.getTypeSignature())));
+        ArrayType arrayType = (ArrayType) typeManager.getParameterizedType(StandardTypes.ARRAY, ImmutableList.of(TypeSignatureParameter.of(type.getTypeSignature())));
         checkCondition(canCastFromJson(arrayType), INVALID_CAST_ARGUMENT, "Cannot cast JSON to %s", arrayType);
-        MethodHandle methodHandle = METHOD_HANDLE.bindTo(arrayType);
+
+        BlockBuilderAppender elementAppender = BlockBuilderAppender.createBlockBuilderValueAppender(arrayType.getElementType());
+        MethodHandle methodHandle = METHOD_HANDLE.bindTo(arrayType).bindTo(elementAppender);
         return new ScalarFunctionImplementation(true, ImmutableList.of(false), methodHandle, isDeterministic());
     }
 
     @UsedByGeneratedCode
-    public static Block toArray(Type arrayType, ConnectorSession connectorSession, Slice json)
+    public static Block toArray(ArrayType arrayType, BlockBuilderAppender elementAppender, ConnectorSession connectorSession, Slice json)
     {
-        try {
-            List<?> array = (List<?>) stackRepresentationToObject(connectorSession, json, arrayType);
-            if (array == null) {
+        try (JsonParser jsonParser = createJsonParser(JSON_FACTORY, json)) {
+            jsonParser.nextToken();
```

Comment:
Now that you rewrote it, should I start failing JSON with trailing tokens? Either change the code or add comment/TODO.